crates/rune/src/module.rs

```diff
@@ -555,6 +558,24 @@ where
         })
     }
 
+    /// Register a constructor method for the current type.
+    pub fn constructor<F, A>(self, constructor: F) -> Result<Self, ContextError>
+    where
+        F: Function<A, Plain, Return = T>,
+    {
+        if self.constructor.is_some() {
+            return Err(ContextError::ConstructorConflict {
+                type_info: T::type_info(),
+            });
+        }
+
+        *self.constructor = Some(Arc::new(move |stack, args| {
+            constructor.fn_call(stack, args)
+        }));
```

Comment:
Without digging too deeply, that seems like that should be the responsibility of the compiler, during the assembly stage? I.e. regardless of which source-level order they're being provided the assembler could (and probably should) re-order the arguments provided to the constructor so it matches the native constructor that's being used.

Reply:
Yeah, that's what I mean. Both are viable options but I have a slight preference towards the second since it minimizes memory use.

What you're after then is sorting the values on the stack by some order after they've been evaluated. I think a swap operation should be sufficient to do that. If you maintain an array of the locations the produced values needs to end up at after construction you should be able to generate the necessary sequence of swaps using something like this (offset from the current size of the stack):

```rust
#[derive(Debug)]
struct Value(usize);

fn main() {
    let mut values: [Value; 4] = [Value(1), Value(2), Value(3), Value(4)];
    let mut order: [usize; 4] = [1, 2, 3, 0];

    for n in 0..order.len() {
        loop {
            let e = order[n];

            if e == n {
                break;
            }

            order.swap(e, n);
            values.swap(e, n);
            println!("swap {e}, {n}");
        }
    }

    dbg!(&values[..], &order[..]);
}
```

Note that the swap instruction currently doesn't exist. The closest would be `replace`, which could be removed in favor of a `swap` which takes two arguments.